[FLINK-3074] Add config option to start YARN AM on port range

rmetzger opened the pull request.

tillrohrmann commented on flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationMasterBase.scala:
```diff
@@ -120,16 +122,73 @@ abstract class ApplicationMasterBase {
+        None
+
+      while (startActorResultOption.isEmpty && portsSet.size() > 0) {
+        if(portsSet == null) {
```
If `portsSet == null` here, then it will have caused a NPE one line above.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/util/NetUtils.java:
```diff
@@ -196,4 +198,35 @@ public static String hostAndPortToUrlString(String host, int port) throws Unknow
+	 * @param logger Logger instance
+	 * @return null if no port was available or an allocated socket.
+	 */
+	public static ServerSocket createSocketFromPorts(Set<Integer> portsSet, SocketFactory factory, Logger logger) throws IOException {
```
Why do you pass the logger as a parameter to the method? That way, it is really hard to know from where the actual logging call was executed.

tillrohrmann commented on flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationMasterBase.scala:
```diff
@@ -120,16 +122,73 @@ abstract class ApplicationMasterBase {
+        if(portsSet == null) {
+          throw new RuntimeException(s"Unable to allocate port for ApplicationMaster in " +
+            s"specified port range: $amPortRange. No more ports available")
+        }
```
maybe linebreak after closing curly bracket.

tillrohrmann commented on flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationMasterBase.scala:
```diff
@@ -120,16 +122,73 @@ abstract class ApplicationMasterBase {
+      var startActorResultOption: Option[(ActorSystem, ActorRef, ActorRef, Option[WebMonitor])] =
+        None
+
+      while (startActorResultOption.isEmpty && portsSet.size() > 0) {
```
Why is the `portsSet.size() > 0` condition included in the while loop condition. Does it change over time?

tillrohrmann replied:
We could also do something like

```
def startActorSystem(portsSet: Set[Integer]): (ActorSystem, ActorRef, ActorRef, Option[WebMonitor]) = {
    val availableSocket = NetUtils.createSocketFromPorts(portsSet, new NetUtils.SocketFactory {
      override def createSocket(port: Int): ServerSocket = new ServerSocket(port)}, log)

    // get port as integer and close socket
    val tryPort = if (availableSocket == null) {
      throw new RuntimeException(s"Unable to allocate port for ApplicationMaster in " +
         s"specified port range: $amPortRange ")
    } else {
     val port = availableSocket.getLocalPort
     availableSocket.close()
     port // return for if
    }

    JobManager.startActorSystemAndJobManagerActors(
      config,
      JobManager.CLUSTER,
      streamingMode,
      ownHostname,
      tryPort,
      getJobManagerClass,
      getArchivistClass
    )
}


@tailrec
def retry[T](fn: => T): Try[T] = {
	Try{ fn } match {
	  case Failure(x: BindException) => retry(fn)
	  case f => f
	}
}
```

tillrohrmann commented:
Thanks for your work @rmetzger. I had some minor comments/questions for clarification.

rmetzger commented:
Thank you for the good review
I've addressed all your concerns. I've also changed the method which parses the configuration string to return an iterator instead of a set.

tillrohrmann commented on flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationMasterBase.scala:
```diff
@@ -120,18 +125,76 @@ abstract class ApplicationMasterBase {
+      val portsIterator = NetUtils.getPortRangeFromString(amPortRange)
+
+      // method to start the actor system.
+      def startActorSystem(portsIterator: java.util.Iterator[Integer]): // return type -> next line
```
When you do line breaks in the method header you usually put each parameter in a new line with double indentation and the return type with the colon in a new line with single indentation. Like: 

```
def foo(
    foo: Bar
    bar: Foo)
  : Foobar = {
}
```

tillrohrmann commented on flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationMasterBase.scala:
```diff
@@ -120,18 +125,76 @@ abstract class ApplicationMasterBase {
+      // method to start the actor system.
+      def startActorSystem(portsIterator: java.util.Iterator[Integer]): // return type -> next line
+        (ActorSystem, ActorRef, ActorRef, Option[WebMonitor]) = {
+        val availableSocket = NetUtils.createSocketFromPorts(portsIterator,
```
`portsIterator` should be in a separate line.

tillrohrmann commented on flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationMasterBase.scala:
```diff
@@ -120,18 +125,76 @@ abstract class ApplicationMasterBase {
+      }
+
+      @tailrec
+      def retry[T](fn: => T, stopCond: => Boolean): Try[T] = {
```
Very nice and scalaesque solution :+1:

tillrohrmann commented on flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationMasterBase.scala:
```diff
@@ -120,18 +125,76 @@ abstract class ApplicationMasterBase {
+        throw new RuntimeException("Unable to start actor system", result.failed.get)
+      }
+      // try starting the actor system
+      val (actorSystem, jmActor, archiveActor, webMonitor) = result.get
```
It would be even more scalaesque if you do it here with pattern matching.

```
result match {
  case Success((actorSystem, jmActor, archiveActor, webMonitor)) =>
  case Failure(failure) =>
}
```

tillrohrmann commented:
Only some minor comments left. +1 for merging after they are are addressed or discarded.

rmetzger commented:
Thank you for the review! I've merged it.

uce commented on flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationMasterBase.scala:
```diff
@@ -120,18 +125,79 @@ abstract class ApplicationMasterBase {
+      }
+
+      // try starting the actor system
+      val result = retry(startActorSystem(portsIterator), {portsIterator.hasNext})
```
The stop condition is wrong. You stop when there are more available ports to check, but you should stop when there are none available any more.
